core-rust/state-manager/src/metrics.rs

```diff
@@ -408,3 +440,30 @@ impl MetricLabel for MempoolAddError {
         }
     }
 }
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum VertexPrepareStopReason {
+    Normal,
+    EpochChange,
+    LimitExceeded(VertexLimitsExceeded),
+}
+
+impl MetricLabel for VertexPrepareStopReason {
+    type StringReturnType = &'static str;
+
+    fn prometheus_label_name(&self) -> Self::StringReturnType {
+        match self {
+            VertexPrepareStopReason::Normal => "Normal",
+            VertexPrepareStopReason::EpochChange => "EpochChange",
+            VertexPrepareStopReason::LimitExceeded(limit_exceeded) => match limit_exceeded {
+                VertexLimitsExceeded::TransactionsCount => "TransactionsCountExceeded",
+                VertexLimitsExceeded::TransactionsSize => "TransactionsSizeExceeded",
+                VertexLimitsExceeded::ExecutionCostUnitsConsumed => "ExecutionCostUnitsConsumed",
+                VertexLimitsExceeded::SubstateReadSize => "SubstateReadSizeExceeded",
+                VertexLimitsExceeded::SubstateReadCount => "SubstateReadCountExceeded",
+                VertexLimitsExceeded::SubstateWriteSize => "SubstateWriteSizeExceeded",
+                VertexLimitsExceeded::SubstateWriteCount => "SubstateWriteCountExceeded",
```

Comment:
On reflection, I'd suggest we phrase these as `_LimitReached` instead of `_Exceeded` - as "Exceeded" sounds like an error to me :)